File: Week_02/cpp/src/_116_PopulatingNextRightPointersInEachNode.cpp
#include <vector>
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};

class Solution
{
public:
    Node *connect(Node *root)
    {
        auto start = root;
        while (start && start->left)
        {
            auto curr = start;
            while (curr)
            {
                curr->left->next = curr->right;
                if (curr->next)
                {
                    curr->right->next = curr->next->left;
                }
                curr = curr->next;
            }
            start = start->left;
        }
        return root;
    }
};

File: Week_02/kotlin/src/_116_PopulatingNextRightPointersInEachNode.kt
class _116_ {
    class Node(var `val`: Int) {
        var left: Node? = null
        var right: Node? = null
        var next: Node? = null
    }

    class Solution {
        fun connect(root: Node?): Node? {
            var start = root
            while (start?.left != null) {
                var node = start
                while (node != null) {
                    node.left!!.next = node.right
                    if (node.next != null) {
                        node.right!!.next = node.next!!.left
                    }
                    node = node.next
                }
                start = start.left
            }
            return root
        }
    }

    class Dfs {
        fun connect(root: Node?): Node? {
            root ?: return null
            root.left?.let {
                it.next = root.right
                root.next?.let {
                    root.right!!.next = it.left
                }
                connect(root.left)
                connect(root.right)
            }
            return root
        }
    }
}